### ISSUE TITLE
JodaTimeAndroid.init(this); crashes with FileNotFoundException for res/raw/joda_time_android_tzdata_zoneinfomap

### ISSUE THREAD

**Issue by aviadmini** [dlew/joda-time-android#70]
I'm using joda-time-android library in a project, but I'm getting quite a number of daily crashes of JodaTimeAndroid.init(this); call which is located in Applicaion overrride class in onCreate(*) method. Here's a stack trace:

    Caused by: java.io.FileNotFoundException: res/raw/joda_time_android_tzdata_zoneinfomap
    at android.content.res.AssetManager.openNonAssetNative(Native Method)
    at android.content.res.AssetManager.openNonAsset(AssetManager.java:446)
    at android.content.res.Resources.openRawResource(Resources.java:1158)
    ... 16 more
    android.content.res.Resources$NotFoundException: File res/raw/joda_time_android_tzdata_zoneinfomap from drawable resource ID #0x7f0701f7
    at android.content.res.Resources.openRawResource(Resources.java:1161)
    at android.content.res.Resources.openRawResource(Resources.java:1132)
    at a.a.a.a.e.a(SourceFile:120)
    at a.a.a.a.e.<init>(SourceFile:39)
    at a.a.a.a.a.a(SourceFile:39)
    at com.example.main.AppOverride.onCreate(SourceFile:46)

It's very weird to me because majority of devices work perfectly fine with it, however this issue also occurs on _some_ quite popular devices like Nexus5/LG G3/several Samsung models etc

**Comment by dlew** [comment]
Are you able to reproduce it with your own devices, or are you only seeing crashes reported?

**Comment by aviadmini** [comment]
Unfortunately only seeing crashes reported.

**Comment by dlew** [comment]
There's not much I can do without a reproduction. Maybe it has something to do with obfuscation?

**Comment by dlew** [comment]
I'm closing this until I see a repro.

**Comment by OlekBodashko** [comment]
I hade the same error:

`
Fatal Exception: java.lang.RuntimeException
Unable to create application com.example.app.MyApplication: android.content.res.Resources$NotFoundException: File res/raw/__tzdata_zoneinfomap from drawable resource ID #0x7f0701cb
`

`Caused by java.io.FileNotFoundException
android.content.res.AssetManager.openNonAssetNative (AssetManager.java)
android.content.res.Resources.openRawResource (Resources.java:940)
net.danlew.android.joda.ResourceZoneInfoProvider.a (SourceFile:120)
net.danlew.android.joda.ResourceZoneInfoProvider.<init> (SourceFile:39)
net.danlew.android.joda.JodaTimeAndroid.a (SourceFile:39)
com.example.app.MyApplication: .onCreate (SourceFile:85)
android.app.Instrumentation.callApplicationOnCreate (Instrumentation.java:1000)
android.app.ActivityThread.handleBindApplication (ActivityThread.java:4566)
android.app.ActivityThread.access$1300 (ActivityThread.java:156)
android.app.ActivityThread$H.handleMessage (ActivityThread.java:1400)
android.os.Looper.loop (Looper.java:153)
android.app.ActivityThread.main (ActivityThread.java:5297)
java.lang.reflect.Method.invokeNative (Method.java)
dalvik.system.NativeStart.main (NativeStart.java)
`

I use joda-time v2.7.2.  Please check this line of the code to see what can cause this error:
net.danlew.android.joda.ResourceZoneInfoProvider.a (SourceFile:120)

**Comment by dlew** [comment]
Do you have a way to reproduce it, or are you just seeing it in crash logs?

**Comment by OlekBodashko** [comment]
Just can see it in crash logs. It's a rare crash but it happens.

**Comment by dlew** [comment]
Unfortunately, I can't really do anything about it without a repro. Maybe it's happening on custom ROMs, maybe the user is doing something screwy with the APK, who knows?

The code is pretty simple - it's just trying to open an asset.

**Comment by mekete** [comment]
I also found a related crash report from Crashlytics. It contains a little more info, and it look like  locale related! 

Caused by java.io.FileNotFoundException
res/raw/joda_asia_qatar

-------Stack Trace Deobusficated------
```
android.content.res.AssetManager.openNonAssetNative (AssetManager.java)
android.content.res.AssetManager.openNonAsset (AssetManager.java:480)
android.content.res.ResourcesImpl.openRawResource (ResourcesImpl.java:337)
android.content.res.Resources.openRawResource (Resources.java:1281)
android.content.res.Resources.openRawResource (Resources.java:1226)
net.danlew.android.joda.ResourceZoneInfoProvider.openResource (ResourceZoneInfoProvider.java:119)
net.danlew.android.joda.ResourceZoneInfoProvider.loadZoneData (ResourceZoneInfoProvider.java:133)
net.danlew.android.joda.ResourceZoneInfoProvider.getZone (ResourceZoneInfoProvider.java:61)
org.joda.time.DateTimeZone.forTimeZone (DateTimeZone.java:359)
org.joda.time.DateTimeZone.getNameKey (DateTimeZone.java:4)
org.joda.time.DateTimeZone.getDefault (DateTimeZone.java:162)
org.joda.time.chrono.GregorianChronology.getInstance (GregorianChronology.java:89)
com.shalom.calendar.HomeActivity. (HomeActivity.java:1)
java.lang.Class.newInstance (Class.java)
android.app.Instrumentation.newActivity (Instrumentation.java:1182)
android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2840)
android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:3059)
android.app.ActivityThread.-wrap11 (ActivityThread.java)
android.app.ActivityThread$H.handleMessage (ActivityThread.java:1724)
android.os.Handler.dispatchMessage (Handler.java:106)
android.os.Looper.loop (Looper.java:164)
android.app.ActivityThread.main (ActivityThread.java:7000)
java.lang.reflect.Method.invoke (Method.java)
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:441)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1408)
```


-------Device Info --------
Device
Model: Samsung Galaxy J7 Prime
Orientation: Portrait
RAM free: 454.89 MB
Disk free: 1.01 GB
Version: 8.1.0 
Rooted: No

Crashed
Date: May 31, 2019, 6:44:00 PM
Joda version 2.9.9

**Comment by dlew** [comment]
That looks more like joda-time-android is being used before being initialized.

**Comment by mekete** [comment]
@dlew JodaAndroid is initialized on ApplicationManager, before anything else. But, there is a situation where the app locale is changed pro-grammatically. When the locale is changed, the app is made to re-start.

**Comment by erikhuizinga** [comment]
TL;DR: I have no way to reproduce, but I experience this bug too.

---

I have this issue too for resource ID `0x7f10016d`. 2 devices on Android Oreo crash. These crashes happens on a `DateTime.now()` call.

The last JTA line that ran is [this](https://github.com/dlew/joda-time-android/blob/5fee3bff3591ee2e107c3779c47ad6228be9ba2a/library/src/main/java/net/danlew/android/joda/ResourceZoneInfoProvider.java#L119).

I think what might be (part of) the problem is that the resource that JTA is looking for might not be bundled correctly (Res ID overwritten? Or not included at all in the AAR compilation process? Or is this a Joda Time resource that has gone missing or was never there?). [This line](https://github.com/dlew/joda-time-android/blob/5fee3bff3591ee2e107c3779c47ad6228be9ba2a/library/src/main/java/net/danlew/android/joda/ResourceZoneInfoProvider.java#L119) is reached, so we know the `resId` is not `0` and therefore probably exists for the requested time zone. The corresponding actual resource file just seems to be missing.

My stacktrace then ends up in `AssetManager#openNonAssetNative` and thinks it's loading a _drawable_ (!):

```
Fatal Exception: android.content.res.Resources$NotFoundException
File from drawable resource ID #0x7f10016d
```

Maybe it thinks it's loading a drawable, because it cannot find any raw resource with `resId` and then `AssetManager` defaults to looking for a nonexistent drawable?

**Comment by dlew** [comment]
I appreciate the extra info, but still, without a repro I'm really not sure what to do. This still strikes me as something to do with a custom OS, because it's hard for me to wrap my head around why getting assets works almost all of the time, but not for a few devices.
